File: 0046-permutations/0046-permutations.py
class Solution:
    def _backtrack(self, nums: List[int], path: List[int], picked: List[bool]) -> None:
        if len(path) == len(nums):
            self.result.append(path[:])
            return
        
        for i in range(len(nums)):
            if not picked[i]:
                path.append(nums[i])
                picked[i] = True
                self._backtrack(nums, path, picked)
                path.pop()  # Backtracking
                picked[i] = False
                
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.result = []           
        self._backtrack(nums, [], [False] * len(nums))
        return self.result

File: 0046-permutations/0046-permutations.cpp
class Solution {
    vector<vector<int>> res;
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> path;
        vector<bool> picked(nums.size(), false);
        backtrack(nums, path, picked);
        return res;
    }
    
    void backtrack(vector<int>& nums, vector<int>& path, vector<bool>& picked) {
        if (path.size() == nums.size()) {
            res.push_back(path);
            return;
        }
        
        for (int i = 0 ; i < nums.size(); i++) {
            if (!picked[i]) {
                path.push_back(nums[i]);
                picked[i] = true;
                backtrack(nums, path, picked);
                path.pop_back();  // Backtracking
                picked[i] = false;
            }
        }
    }
};